tests/unit/executors/indexers/test_numpyindexer.py

```diff
@@ -35,13 +35,76 @@ def test_numpy_indexer(batch_size, compress_level, test_metas):
         assert idx.shape == (num_query, 4)
 
 
+@pytest.mark.parametrize('batch_size, compress_level', [(None, 0), (None, 1), (2, 0), (2, 1)])
+def test_numpy_update_delete(batch_size, compress_level, test_metas):
+    np.random.seed(500)
+    num_dim = 3
+    vec_idx = np.array([12, 112, 903])
+    vec = np.random.random([len(vec_idx), num_dim])
+
+    with NumpyIndexer(metric='euclidean', index_filename='np.test.gz', compress_level=compress_level,
+                      metas=test_metas) as indexer:
+        indexer.add(vec_idx, vec)
+        indexer.save()
+        assert indexer.num_dim == num_dim
+        assert indexer.size == len(vec_idx)
+        assert Path(indexer.index_abspath).exists()
+        save_abspath = indexer.save_abspath
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        query_results = indexer.query_by_id(vec_idx)
+        assert np.array_equal(vec, query_results)
+
+    # update
+    key_to_update = vec_idx[0]
+    data_to_update = np.random.random([1, num_dim])
+    # nonexistent key
+    random_keys = [999]
+    random_data = np.random.random([1, num_dim])
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        # this will fail cause the key doesn't exist
+        with pytest.raises(KeyError):
+            indexer.update(random_keys, random_data)
+        indexer.update([key_to_update], data_to_update)
+        indexer.save()
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        query_results = indexer.query_by_id([key_to_update])
+        assert np.array_equal(data_to_update, query_results)
+
+    # delete
+    keys_to_delete = 1
+    vec_idx_to_delete = vec_idx[:keys_to_delete]
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        indexer.delete(vec_idx_to_delete)
+        indexer.save()
+        assert indexer.size == len(vec_idx) - keys_to_delete
+
+    assert indexer.size == len(vec_idx) - keys_to_delete
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        assert indexer.size == len(vec_idx) - keys_to_delete
+        # this will fail. key doesn't exist
+        with pytest.raises(KeyError):
+            _ = indexer.query_by_id(vec_idx)
+        query_results = indexer.query_by_id(vec_idx[keys_to_delete:])
+        expected = vec[keys_to_delete:]
+        np.testing.assert_allclose(query_results, expected, equal_nan=True)
+
+
 @pytest.mark.parametrize('batch_size, compress_level', [(None, 0), (None, 1), (16, 0), (16, 1)])
-def test_numpy_indexer_known(batch_size, compress_level, test_metas):
+def test_numpy_indexer_known_and_delete(batch_size, compress_level, test_metas):
```

Comment:
Is it possible to duplicate this test instead of changing it? This test is done for every `indexer` in the hub and I would prefer not to break this.

Reply:
Sure, done